examples/darwin-framework-tool/commands/provider/OTAProviderDelegate.mm

```diff
@@ -0,0 +1,173 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "OTAProviderDelegate.h"
+#import <Matter/Matter.h>
+
+constexpr uint8_t kUpdateTokenLen = 32;
+
+@interface OTAProviderDelegate ()
+@property NSString * mOTAFilePath;
+@property DeviceSoftwareVersionModel * candidate;
+@end
+
+@implementation OTAProviderDelegate
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        _nodeID = @(0);
+        _selectedCandidate = [[DeviceSoftwareVersionModel alloc] init];
+        _userConsentState = OTAProviderUserUnknown;
+    }
+    return self;
+}
+
+- (void)handleQueryImage:(MTROtaSoftwareUpdateProviderClusterQueryImageParams * _Nonnull)params
+       completionHandler:(void (^_Nonnull)(MTROtaSoftwareUpdateProviderClusterQueryImageResponseParams * _Nullable data,
+                             NSError * _Nullable error))completionHandler
+{
+    NSError * error;
+    _selectedCandidate.status = @(MTROtaSoftwareUpdateProviderOTAQueryStatusNotAvailable);
+
+    if ([self SelectOTACandidate:params.vendorId rPID:params.productId rSV:params.softwareVersion]) {
+        _selectedCandidate.status = @(MTROtaSoftwareUpdateProviderOTAQueryStatusUpdateAvailable);
+        _selectedCandidate.updateToken = [self generateUpdateToken];
+        if (params.requestorCanConsent.integerValue == 1) {
+            _selectedCandidate.userConsentNeeded
+                = (_userConsentState == OTAProviderUserUnknown || _userConsentState == OTAProviderUserDenied) ? @(1) : @(0);
+            NSLog(@"User Consent Needed: %@", _selectedCandidate.userConsentNeeded);
+        } else {
+            NSLog(@"Requestor cannot obtain user consent. Our State: %hhu", _userConsentState);
+            switch (_userConsentState) {
+            case OTAProviderUserGranted:
+                NSLog(@"User Consent Granted");
+                _queryImageStatus = MTROtaSoftwareUpdateProviderOTAQueryStatusUpdateAvailable;
```

Comment:
Maybe leave a TODO somewhere to start BDX when that is exposed to Darwin framework?

Reply:
Added.